python/ray/ml/train/integrations/sklearn/sklearn_trainer.py

```diff
@@ -112,26 +152,272 @@ def __init__(
         datasets: Dict[str, GenDataset],
         label_column: Optional[str] = None,
+        params: Optional[Dict[str, Any]] = None,
         scoring: Optional[ScoringType] = None,
         cv: Optional[CVType] = None,
+        parallelise_cv: Optional[bool] = None,
+        return_train_score_cv: bool = False,
         scaling_config: Optional[ScalingConfig] = None,
         run_config: Optional[RunConfig] = None,
         preprocessor: Optional[Preprocessor] = None,
-        resume_from_checkpoint: Optional[Checkpoint] = None,
-        **fit_params
+        **fit_params,
     ):
-        self.estimator = estimator
+        self.estimator = clone(estimator)
         self.label_column = label_column
+        self.params = params or {}
         self.fit_params = fit_params
         self.scoring = scoring
         self.cv = cv
+        self.parallelise_cv = parallelise_cv
+        self.return_train_score_cv = return_train_score_cv
         super().__init__(
             scaling_config=scaling_config,
             run_config=run_config,
             datasets=datasets,
             preprocessor=preprocessor,
-            resume_from_checkpoint=resume_from_checkpoint,
+            resume_from_checkpoint=None,
         )
 
+    def _validate_attributes(self):
+        super()._validate_attributes()
+        if TRAIN_DATASET_KEY not in self.datasets:
+            raise KeyError(
+                f"'{TRAIN_DATASET_KEY}' key must be preset in `datasets`. "
+                f"Got {list(self.datasets.keys())}"
+            )
+        if "cv" in self.datasets:
+            raise KeyError(
+                "'cv' is a reserved key. Please choose a different key "
+                "for the dataset."
+            )
+        if (
+            not isinstance(self.parallelise_cv, bool)
+            and self.parallelise_cv is not None
+        ):
+            raise ValueError(
+                "`parallelise_cv` must be a bool or None, got "
+                f"'{self.parallelise_cv}'"
+            )
+        scaling_config = ScalingConfigDataClass(**self.scaling_config)
+        ScalingConfigDataClass.validate_config(
+            config=scaling_config,
+            allowed_keys=["trainer_resources"],
+            scaling_config_arg_name="scaling_config",
+            caller_name=self.__class__.__name__,
+        )
+        if (
+            self.cv
+            and self.parallelise_cv
+            and scaling_config.trainer_resources.get("GPU", 0)
+        ):
+            raise ValueError(
+                "`parallelise_cv` cannot be True if there are GPUs assigned to the "
+                "trainer."
+            )
+
+    def _set_estimator_cpus(self, estimator: BaseEstimator, num_cpus: int) -> bool:
+        """Returns True if there are any params related to CPUs."""
+        cpu_params = {
+            param: num_cpus
+            for param in estimator.get_params(deep=True)
+            if param.endswith("n_jobs") or param.endswith("thread_count")
+        }
```

Comment:
How many of these parameter do we usually have? Aren't we overassigning CPUs here if we have more than one?

Reply:
Usually we will only have one `n_jobs`. However, in a case a user passes a sklearn Pipeline, each element of it may have its own `n_jobs` (`element__n_jobs`). Pipelines are ran sequentially, so there isn't a real possibility of overassignment.

Reply:
Made the behavior toggelable